### PR TITLE
Change external behaviour to work with hardware-specific types

### PR DESCRIPTION
**Pull request by guilhermeleobas**
As title

### CODE REVIEW

File: rbc/external.py [hunk 1]
```diff
@@ -16,55 +17,64 @@ def fromobject(cls, signature, name: str = None):
...
         """
         # Make inner function for the actual work
-        t = Type.fromobject(signature)
-        if not t.is_function:
-            raise ValueError("signature must represent a function type")
+        target_info = TargetInfo.host()
+        with target_info:
+            t = Type.fromobject(signature)
+            if not t.is_function:
+                raise ValueError("signature must represent a function type")
```
[the remark this answers is not in the document]
Comment (pearu) [reply rbc/external.py:24]: We should merge https://github.com/xnd-project/rbc/pull/274 first, and then use
```
with TargetInfo.dummy():
    if not Type.fromobject(signature).is_function:
        raise ValueError("signature must represent a function type")
```

File: rbc/external.py [hunk 2]
```diff
@@ -16,55 +17,64 @@ def fromobject(cls, signature, name: str = None):
...
+            if name is None:
+                name = t.name
+            if not name:
+                raise ValueError(
+                    f"external function name not specified for signature {signature}"
+                )
 
-        return cls(name, t.tonumba())
+            typ = t
+            t = t.tonumba()
```
Comment (pearu) [inline rbc/external.py:34]: We should not materialize the signature here!

File: rbc/external.py [hunk 3]
```diff
@@ -16,55 +17,64 @@ def fromobject(cls, signature, name: str = None):
...
         Parameters
         ----------
         name : str
             The name of the external function
         signature : Numba function signature
             A numba function type signature. i.e. (float64, float64) -> int64
+        rbc_type: RBC typesystem Type
+            A type from the RBC typesystem type
         """
         self._signature = signature
```
Comment (pearu) [inline rbc/external.py:49]: We should eliminate `_signature` that is numba function type.

Comment (pearu) [reply rbc/external.py:49]: And rename `rbc_type` -> `signature`

**Review by pearu — CHANGES_REQUESTED** [review]
Looks good.

`External` properties `return_type`, `args` are not used and should be deleted.

File: rbc/external.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply rbc/external.py:24]: > This basically checks that signature represents a function

Not just that, but to parse the signature and get the function name if `name` is not specified as a keyword arg

File: rbc/external.py [hunk 4]
```diff
@@ -44,27 +47,28 @@ def __init__(self, name: str, signature: types.FunctionType):
...
+                t = Type.fromobject(self.obj._signature).tonumba()
 
-        # lowering
-        def codegen(context, builder, sig, args):
-            fndesc = funcdesc.ExternalFunctionDescriptor(
-                self.name, sig.return_type, sig.args
-            )
-            func = context.declare_external_function(builder.module, fndesc)
-            return builder.call(func, args)
+                name = self.key
```
Comment (pearu) [inline rbc/external.py:57]: When external is going to support multiple signatures, as in `external(['double log2(double)', 'float log2f(float)'], name='log2')`, then we should likely use:
```
name = t.name or self.key
```

File: rbc/external.py [hunk 5]
```diff
@@ -44,27 +47,28 @@ def __init__(self, name: str, signature: types.FunctionType):
...
+            obj = self
+            key = self.name
 
-        template = make_concrete_template(
-            self.name, key=self.name, signatures=[self.signature]
-        )
-        infer(template)
-        infer_global(self, types.Function(template))
+            def generic(self, args, kws):
+                t = Type.fromobject(self.obj._signature).tonumba()
```
Comment (pearu) [inline rbc/external.py:55]: I presume that in multiple signature cases (see below), this is the place where `t` needs to depend on `args`.

**Review by pearu — APPROVED** [review]
LGTM.

Added few comments for future extensions.

File: rbc/external.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pearu) [reply rbc/external.py]: Right. In future, when the signatures need to contain the function names, we may want to eliminate the `name` argument.
So, let's not use `name` too much :)

[the remark this answers is not in the document]
Comment (guilhermeleobas) [reply rbc/external.py]: The name might be useful to use as a key of `AbstractTemplate`.

File: rbc/external.py (hunk 5 above)
Comment (guilhermeleobas) [reply rbc/external.py:55]: exactly

**Pull request merged**